routes/api/files/download-directory.tsx

```diff
@@ -0,0 +1,241 @@
+import { Handlers } from 'fresh/server.ts';
+import { join, resolve } from '@std/path';
+
+import { FreshContextState } from '/lib/types.ts';
+import { AppConfig } from '/lib/config.ts';
+import { DirectoryModel, FileModel } from '/lib/models/files.ts';
+
+interface Data {}
+
+export const handler: Handlers<Data, FreshContextState> = {
+  async GET(request, context) {
+    if (!context.state.user) {
+      return new Response('Redirect', { status: 303, headers: { 'Location': `/login` } });
+    }
+
+    const config = await AppConfig.getConfig();
+
+    // Check if directory downloads are enabled
+    if (!config.files?.allowDirectoryDownloads) {
+      return new Response('Directory downloads are not enabled', { status: 403 });
+    }
+
+    const searchParams = new URL(request.url).searchParams;
+    let directoryPath = searchParams.get('path') || '/';
+
+    // Send invalid paths back to root
+    if (!directoryPath.startsWith('/') || directoryPath.includes('../')) {
+      return new Response('Invalid path', { status: 400 });
+    }
+
+    // Always append a trailing slash
+    if (!directoryPath.endsWith('/')) {
+      directoryPath = `${directoryPath}/`;
+    }
+
+    try {
+      // Get all files and subdirectories recursively
```

Comment:
Instead of all of this and a kind-of-library, I'd rather just call `zip path` and ensure it's installed in the `Dockerfile` (and perhaps mention in the README as a requirement).

Also, before all this, call `ensureUserPathIsValidAndSecurelyAccessible` on the requested path.

Reply:
Can do. I wasn't sure what is and isn't allowed with your no-dependency rule so I did my best attempt at something without one =)